Fix nested generic typerefs applying generic params at the wrong level

augusto2112 opened the pull request:
Generic params of typerefs are supposed to be "attached" on the level
they belong, not as a flat list, unlike other parts of the system. Fix
the application of bound generic params by checking how many were
already applied in the hierarchy and ignoring those already attached.

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2932,34 +2915,64 @@ class MetadataReader {
+    if (auto parentDescriptor = readParentContextDescriptor(descriptor))
+      if (parentDescriptor->isResolved())
+        if (auto resolvedParent = parentDescriptor->getResolved())
+          parent = readNominalTypeFromDescriptor(metadata, resolvedParent);
```
Not sure if this is ok, I'm passing the metadata of the current descriptor, not the parent.

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test

mikeash commented on include/swift/Reflection/TypeRef.h:
```diff
@@ -307,10 +325,17 @@ class BoundGenericTypeRef final : public TypeRef, public NominalTypeTrait {
+    if (Parent)
+      GenericParams.erase(GenericParams.begin(),
+                          GenericParams.begin() +
+                              GetNumberOfGenericParametersInHierarchy(Parent));
```
Can you add a bounds check here to fail gracefully if `GetNumberOfGenericParametersInHierarchy(Parent) > GenericParams.size()`? We sometimes get weird values due to corrupt or misidentified data in the target process.

mikeash reviewed (approved):
One minor request, LGTM.

adrian-prantl commented on include/swift/Reflection/TypeRef.h:
```diff
@@ -283,6 +283,24 @@ class NominalTypeRef final : public TypeRef, public NominalTypeTrait {
 class BoundGenericTypeRef final : public TypeRef, public NominalTypeTrait {
   std::vector<const TypeRef *> GenericParams;
 
+  static uint64_t
```
Doxygen comment?

adrian-prantl replied:
(What's not obvious from the name is that this counts the generic parameters in the parent only.)

adrian-prantl reviewed (commented):
Nice!

augusto2112 commented:
@swift-ci smoke test

jckarter commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2905,25 +2909,50 @@ class MetadataReader {
+  /// single flat array.
+  template <typename T = BuilderType,
+            std::enable_if_t<std::is_same<T, reflection::TypeRefBuilder>::value,
+                             bool> = true>
```
Instead of hardcoding `TypeRefBuilder` here, it might be cleaner to look for a new trait on `BuilderType`, so we can do `enable_if_t<BuilderType::needsToPrecomputeParentGenericContextShapes>` or something like that.

augusto2112 committed:
Fix nested generic typerefs applying generic params at the wrong level

Generic params of typerefs are supposed to be "attached" on the level
they belong, not as a flat list, unlike other parts of the system. Fix
the application of bound generic params by checking how many were
already applied in the hierarchy and ignoring those already attached.

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@jckarter updated

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on include/swift/AST/ASTDemangler.h:
```diff
@@ -64,6 +64,8 @@ class ASTBuilder {
   using BuiltRequirement = swift::Requirement;
   using BuiltSubstitutionMap = swift::SubstitutionMap;
 
+  static constexpr bool needsToPrecomputeParentGenericContextShapes = false;
```
Should we Doxygen-document what this parameter means?

adrian-prantl commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2820,6 +2827,41 @@ class MetadataReader {
+    auto node = buildContextMangling(descriptor, dem);
+    if (!node || node->getKind() != Node::Kind::Type)
+      return BuiltTypeDecl();
+    std::vector<size_t> paramsPerLevel;
```
`llvm::SmallVector<size_t, 4>` ?

adrian-prantl commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2820,6 +2827,41 @@ class MetadataReader {
+      return BuiltTypeDecl();
+    std::vector<size_t> paramsPerLevel;
+    size_t runningCount = 0;
+    std::function<void(ContextDescriptorRef current, size_t &)> countLevels =
```
Comment what the lambda does?

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
Seems like this is failing for some other reason:

```
error: terminated(1): /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/swiftpm-xctest-helper /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.xctest /var/folders/pf/p5g2vb2j0yx1x6vlmxdppc700000gn/T/TemporaryFile.ke3aih output:
    error: unableToLoadBundle("/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.xctest")
    2022-06-23 00:26:01.911 swiftpm-xctest-helper[70753:3102817] Error loading /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.xctest/Contents/MacOS/SwiftPMPackageTests:  dlopen(/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.xctest/Contents/MacOS/SwiftPMPackageTests, 265): Library not loaded: @rpath/libswift_Concurrency.dylib
      Referenced from: /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/release/SwiftPMPackageTests.xctest/Contents/MacOS/SwiftPMPackageTests
      Reason: image not found
```

augusto2112 commented:
@swift-ci smoke test
